AnnoDesigner/AnnoCanvas.xaml.cs

```diff
@@ -29,6 +29,261 @@
+    {
+        private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+
+        private double _delta_x;
+        private double _delta_y;
+
+        public MoveAllObjects(double delta_x, double delta_y)
+        {
+            _delta_x = delta_x;
+            _delta_y = delta_y; 
+        }
+
+        public void PerformAction(List<LayoutObject> placedObjects)
+        {
+            logger.Debug("Perform move all action.");
+            foreach (var curLayoutObject in placedObjects)
+            {
+                curLayoutObject.Position = new Point(curLayoutObject.Position.X + _delta_x, curLayoutObject.Position.Y + _delta_y);
+            }
+        }
+
+        public void UndoAction(List<LayoutObject> placedObjects)
+        {
+            logger.Debug("Undo move all action.");
+            foreach (var curLayoutObject in placedObjects)
+            {
+                curLayoutObject.Position = new Point(curLayoutObject.Position.X - _delta_x, curLayoutObject.Position.Y - _delta_y);
+            }
+        }
+
+        public bool CombinableWith(ICombinableAction other)
+        {
+            if (other is MoveAllObjects)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        public void CombineActions(ICombinableAction other)
+        {
+            logger.Debug("Combine two move all actions.");
+            if (!CombinableWith(other)) 
+            {
+                // If this has happened then we are going to end up with some screwed up action history
+                throw new System.ArgumentException("Cannot combine these ICombinableActions", "toCombine");
+            }
+
+            MoveAllObjects toCombine = (MoveAllObjects)other;
+
+            _delta_x += toCombine._delta_x;
+            _delta_y += toCombine._delta_y;
+        }
+    }
+
+    /// <summary>
+    /// Doesn't do anything, it is used to split up actions so they aren't combined together into a single action.
+    ///
+    /// For example multiple consecutive MoveAllObjects actions will be combined into a single one (to save memory and
+    /// so the user doesn't need to undo them all individually). However, we do want some move actions to remain separated
+    /// these should be separated by a NoOpAction.
+    /// </summary>
+    public class NoOpAction : IAction 
+    {
+        private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+
+        public NoOpAction() {}
+
+
+        public void PerformAction(List<LayoutObject> placedObjects)
+        {
+            logger.Debug("Performing no-op action.");
+        }
+
+        public void UndoAction(List<LayoutObject> placedObjects)
+        {
+            logger.Debug("Undoing no-op action");
+        }
+    }
+
+    /// <summary>
+    /// Class which manages the actions that have been taken. Allows performing new actions, and undoing and redoing actions.
+    /// </summary>
+    public class ActionManager
```

Comment:
Please move these undo related classes to some other file, so it is clear which class are you modifying in partial diffs.